project/ian-chiu/README.md

```diff
@@ -0,0 +1,185 @@
+# Celluar Automata 3D
+
+## Basic Information
+
+A software to simulate and visualize 3D cellular automata
+* github repository: [https://github.com/ian-chiu/celluar-automata-3d](https://github.com/ian-chiu/celluar-automata-3d)
+
+## Problem To Solve
+
+Like the 2D Cellular Automata (Conway's Game of Life), the goal is to simulate and visualize cellular automata but extend from 2D to 3D.
+
+| ![Conway's Game of Life](https://upload.wikimedia.org/wikipedia/commons/e/e5/Gospers_glider_gun.gif) | ![](https://i.imgur.com/nQ7m3OU.png) |
+|:--------:|:--------:|
+| **2D Cellular Automata - Conway's Game of Life** | **3D Cellular Automata - from Softology's Blog** |
+
+Cellular Automata Definition from Tech Target:
+> A cellular automation is a collection of cells arranged in a grid of specified shape, such that each cell changes state as a function of time, according to a defined set of rules driven by the states of neighboring cells.
+
+### Cell Rules
+
+Cell rules can be represented by: 
+**Survival / Spawn / State / Neighbor**
+
+#### Survival
+
+A list of acceptable numbers of neighbors that are alive (in state 1) to keep a cell survived (stay in state 1). If the number does not match any in the list, the cell start dying (its state start increasing every tick). If a cell's state is greater than its maximum state, it died (back to state 0).
+
+#### Spawn
+
+A list of acceptable numbers of neighbors that are alive (in state 1) to spawn a new cell. If a cell is empty (in state 0) and the number matches any in the list, the cell get spawned (go from state 0 to 1).
+
+#### State
+
+The maximum number of state a cell can have. If a cell's state is greater this number, then it died (back to state 0).
+
+#### Neighbor
+
+There are two kinds of neighbor, Von Neumann(VN) and Moore(M). Von Neumann is like a rubik cube, the cell at the center has 26 neighbors. Moore has only 4 neighbors, top, down, front and back.
+
+### Optimization
+
+#### Multithreading and SIMD... or any parallel computing techinques
+
+Use parallel computing techinques, like multithreading and SIMD operations, to increase computation speed of calculating all cell states in the next frame under a rule. 
+
+If a grid is 100x100x100, we have to calculate next states of 1000000 cells according to their neighbors every frame, which needs a lot of computation.
+
+#### Greedy Meshing Algorithm
+
+The idea of the algorithm is combining adjacent blocks to reduce the part count while visually keeping everything the same, which is useful to reduce draw calls. 
+
+For a 100x100x100 grid, we have to draw maximum 1000000 cubes if all cells are not empty. If the number of draw calls is too large, the communication between cpu and gpu will be overwhelming, which can greatly reduce the performance of rendering.
+
+![](https://i.imgur.com/SCIFdNj.png)
+
+## Prospective Users
```

Comment:
It's a comprehensive survey.  I like it.

Reply:
Thanks.